MAINT: A few updates to the array_api

asmeurer opened the pull request:
A few minor updates to the array_api namespace based on recent updates to the spec. 

CC @seberg you may actually want to review this one before it is merged this time.

asmeurer committed:
Allow casting in the array API asarray()

asmeurer committed:
Restrict multidimensional indexing in the array API namespace

The spec has recently been updated to only require multiaxis (i.e., tuple)
indices in the case where every axis is indexed, meaning there are either as
many indices as axes or the index has an ellipsis.

seberg commented:
The main comment I remember was that you had a lot of complexity to support `arr[boolean_arr,]` when maybe all you need to support is `arr[boolean_arr]`.  But I guess there was no change/decision w.r.t. to that one made.  And isn't the most important point the non-multi-dim index part? `arr3d[0]` not being strictly defined or did you decide against that after all?

mattip reviewed (commented):
linting

asmeurer committed:
Fix type promotion for numpy.array_api.where

where does value-based promotion for 0-dimensional arrays, so we use the same
trick as in the Array operators to avoid this.

asmeurer committed:
Print empty array_api arrays using empty()

Printing behavior isn't required by the spec. This is just to make things
easier to understand, especially with the array API test suite.

asmeurer committed:
Fix an incorrect slice bounds guard in the array API

asmeurer committed:
Disallow multiple different dtypes in the input to np.array_api.meshgrid

asmeurer committed:
Remove DLPack support from numpy.array_api.asarray()

from_dlpack() should be used to create arrays using DLPack.

asmeurer committed:
Remove __len__ from the array API array object

asmeurer committed:
Add astype() to numpy.array_api

asmeurer committed:
Update the unique_* functions in numpy.array_api

unique() in the array API was replaced with three separate functions,
unique_all(), unique_inverse(), and unique_values(), in order to avoid
polymorphic return types.

Additionally, it should be noted that these functions to not currently conform
to the spec with respect to NaN behavior. The spec requires multiple NaNs to
be returned, but np.unique() returns a single NaN. Since this is currently an
open issue in NumPy to possibly revert, I have not yet worked around this. See
https://github.com/numpy/numpy/issues/20326.

asmeurer committed:
Add the stream argument to the array API to_device method

This does nothing in NumPy, and is just present so that the signature is valid
according to the spec.

asmeurer committed:
Use the NamedTuple classes for the type signatures

asmeurer committed:
Add unique_counts to the array API namespace

asmeurer committed:
Remove some unused imports

asmeurer committed:
Update the array_api indexing restrictions

The "multiaxis indexing must index every axis explicitly or use an ellipsis"
was supposed to include any type of index, not just tuple indices.

asmeurer committed:
Merge branch 'main' into array-api-updates2

asmeurer commented:
This is ready for review. There have been several small changes to the spec since the last time anyone has looked at this. With this PR, everything should now be up-to-date, except for a handful of things which are being updated in separate pull requests.

rgommers commented:
@asmeurer there is one test failure, `test_device_property`. Could you fix that?

rgommers commented:
The boolean indexing question higher up (https://github.com/numpy/numpy/pull/20066#issuecomment-938177737) also deserves a reply.

rgommers commented on numpy/array_api/_set_functions.py:
```diff
@@ -2,30 +2,91 @@
 
 import numpy as np
 
+# Note: np.unique() is split into fout functions in the array API:
```
minor: typo in `fout`

rgommers commented on numpy/array_api/tests/test_creation_functions.py:
```diff
@@ -124,3 +126,11 @@ def test_zeros_like_errors():
+
+def test_meshgrid_dtype_errors():
+    # Doesn't raise
+    meshgrid()
```
A bit surprising:
```
>>> np.meshgrid()
[]
```
But ah well, that is what it is.

asmeurer replied:
I don't see why that's unexpected. `meshgrid(*args)` should return a list of the same length as `args`. I don't know if it's ever useful. Possibly if there is code that passes an unknown set of arrays fo meshgrid it can be the correct null condition.

More importantly though I think if functions with `*args` can produce meaningful and consistent results with empty `args`, even if the result is itself something trivial like an empty list, that is preferable to raising an exception. That way the function actually matches its signature. Consider something like `result_type(*arrays_and_dtypes)` which is a bit misleading because `result_type()` does not actually work.

rgommers reviewed (approved):
This all LGTM modulo the test failure.

asmeurer commented:
> @asmeurer there is one test failure, test_device_property. Could you fix that?

Sorry about that. https://github.com/numpy/numpy/issues/20335 has messed up my ability to run the tests locally.

BvB93 commented on numpy/array_api/_array_object.py:
```diff
@@ -995,7 +1009,9 @@ def __rxor__(self: Array, other: Union[int, bool, Array], /) -> Array:
         return self.__class__._new(res)
 
-    def to_device(self: Array, device: Device, /) -> Array:
+    def to_device(self: Array, device: Device, /, stream: Optional[Union[int, Any]] = None) -> Array:
```
```suggestion
    def to_device(self: Array, device: Device, /, stream: None = None) -> Array:
```
Considering numpy's array-api implementation only accepts `None` as stream-value.

leofang replied:
It would be nice if NumPy could keep the signatures as prescribed in the spec, otherwise we will have to fix it downstream... It's simple enough to reject any non-None objects here, right?

BvB93 replied:
Can you elaborate why this would not be up to spec? The `__dlpack__` [docs](https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-self-stream-none) explicitly state that devices lacking stream-support should only accept `None` for the `stream` parameter.

leofang replied:
For NumPy it's ok. For CuPy we will have to fix it when we apply the patch from this PR. It's an extra step that I have to note to remind myself 😥

asmeurer commented:
> The main comment I remember was that you had a lot of complexity to support arr[boolean_arr,] when maybe all you need to support is arr[boolean_arr]. But I guess there was no change/decision w.r.t. to that one made. And isn't the most important point the non-multi-dim index part? arr3d[0] not being strictly defined or did you decide against that after all?

I did address the point of treating `a[0]` and `a[0,]` differently. That was just a misreading of the spec on my part. I think we could drop `a[boolean_arr,]` if that simplifies the code. What was the suggested way to do that? Any refactoring of the index validation should be fine so long as it passes the tests in test_array_object.py as well as the test_indexing test in the array API testsuite.

asmeurer committed:
Use a simpler type annotation for the array API to_device method

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

asmeurer committed:
Fix a typo

asmeurer committed:
Fix a test failure in the array_api submodule

The array_api cannot use the NumPy testing functions because array_api arrays
do not mix with NumPy arrays, and also NumPy testing functions may use APIs
that aren't supported in the array API.

asmeurer committed:
Merge branch 'array-api-updates2' of github.com:asmeurer/numpy into array-api-updates2

asmeurer committed:
Merge branch 'main' into array-api-updates2

asmeurer committed:
Add dlpack support to the array_api submodule

rgommers commented:
This all looks good. CI failures are okay: linter ones are uninteresting, the typing error was fixed in `main` yesterday, and the segfault is unrelated (happening in other PRs). So in this goes.

> I did address the point of treating `a[0]` and `a[0,]` differently. That was just a misreading of the spec on my part. I think we could drop `a[boolean_arr,]` if that simplifies the code. What was the suggested way to do that?

Seems fine either way to me, and can be done in a separate follow-up given that it's a discussion that predates this PR. Let's see if @seberg has a preference here.

rgommers committed:
Merge branch 'main' into array-api-updates2

rgommers commented:
Ah wait, there's even more failures. They all look unrelated, but rerunning CI anyway to get this a bit more green.

rgommers commented:
Looking much better now, in it goes. Thanks @asmeurer & reviewers!

seberg commented:
A followup is fine and would be nice.  I don't recall the logic exactly, what I do remember is that the indexing check was very complex (with some recursive calls), and a lot of that complexity was there to accomodate `arr[boolean_arr,]`.  I believe in the original PR I had proposed a simplification.
That said, I think it is correct, just seemed more complex than necessary.

seberg commented:
PS: And since this is experimental and a corner case, I don't think it would matter if you change `arr[boolean_arr,]` from being allowed to being rejected.
